### PR TITLE
Implement typeid(struct/class).name for CTFE

### PR DESCRIPTION
**Pull request by UplinkCoder**
This implements a small subset {.name} of typeid for CTFE.

I think this is a small but valuable addition.

### CODE REVIEW

**Comment by WalterBright** [comment]
Isn't that what `.stringof` is for?

**Comment by UplinkCoder** [comment]
this is supposed to return the dynamic type name.
if you replace the `typeid(o).name` with `typeof(o).stringof` you would get the static type.
@WalterBright

**Comment by schveiguy** [comment]
Ideally, all items of a `typeid` generated by the compiler would be accessible. It's understandable that not all this is available, since TypeInfo is part compiler-generated and part library.

I would like to see `initializer` as well.

File: src/dmd/dinterpret.d [hunk 1]
```diff
@@ -6095,6 +6095,30 @@ public:
             se = (cast(ClassReferenceExp)ex).value;
             i = (cast(ClassReferenceExp)ex).findFieldIndexByName(v);
         }
+        else if (ex.op == TOK.typeid_)
+        {
+            auto tie = cast(TypeidExp) ex;
+            if (v.ident == Identifier.idPool("name"))
+            {
+                if (auto t = isType(tie.obj))
```
Comment (RazvanN7) [inline src/dmd/dinterpret.d]: `if (auto t = isType((cast(TypeidExp)ex).obj))`

This is the only use of `tie`, so there is no point in declaring a variable.

File: src/dmd/dinterpret.d [hunk 2]
```diff
@@ -6095,6 +6095,30 @@ public:
...
+                {
+                    auto sym = t.toDsymbol(null);
+                    auto ident = (sym ? sym.ident : null);
+                    if (ident)
```
Comment (RazvanN7) [inline src/dmd/dinterpret.d]: `if (auto ident = sym ? sym.ident : null)`

File: src/dmd/dinterpret.d [hunk 3]
```diff
@@ -6095,6 +6095,30 @@ public:
...
+                    {
+                        result = new StringExp(e.loc, ident.toString());
+                        result.expressionSemantic(null);
+                        return ;
+                    }
+                }
+            }
+            // We get here if we could not resolve the name or we were not asked for the name
+            {
```
Comment (RazvanN7) [inline src/dmd/dinterpret.d]: The scope block is unnecessary.

File: src/dmd/dinterpret.d [hunk 4]
```diff
@@ -6095,6 +6095,30 @@ public:
...
+                e.error("only typeid(class/struct).name is implemented at compile time");
```
Comment (RazvanN7) [inline src/dmd/dinterpret.d]: I would rephrase the error message to something along the lines of: "At compile time, `typeid` works only if the provided type is `class` or `struct` and the `name` property is accessed: `typeid(class/struct).name`"

File: test/compilable/typeid_name.d [hunk 5]
```diff
@@ -0,0 +1,14 @@
+
```
Comment (RazvanN7) [inline test/compilable/typeid_name.d]: A `fail_compilation` test should be added to cover the introduced error message.

**Review by RazvanN7 — CHANGES_REQUESTED** [review]
Looks good otherwise.

File: src/dmd/dinterpret.d (hunk 2 above)
Comment (UplinkCoder) [reply src/dmd/dinterpret.d]: I don't want to increase the nesting level unnecessarily.

File: src/dmd/dinterpret.d (hunk 3 above)
Comment (UplinkCoder) [reply src/dmd/dinterpret.d]: The block is the target of multiple failing branches, I prefer the error-code to be grouped that way.
I plan to extend this functionality and that format is the friendly to extension.

File: src/dmd/dinterpret.d (hunk 1 above)
Comment (UplinkCoder) [reply src/dmd/dinterpret.d]: it's in line with how the rest of the code in here does it.
`also tie` will be used more in the future, as I intend to extend  the CTFE typeid support.

File: test/compilable/typeid_name.d (hunk 5 above)
Comment (UplinkCoder) [reply test/compilable/typeid_name.d]: I'd rather not introduce an test for an error message which might have to be removed.
the restriction to `typeid().name` is not intentional. and if I'd have to remove a test when extending it, that would look bad.

File: src/dmd/dinterpret.d (hunk 4 above)
Comment (UplinkCoder) [reply src/dmd/dinterpret.d]: I don't really care.

Comment (schveiguy) [reply src/dmd/dinterpret.d]: `typeid(class/struct).name` is a somewhat ambiguous message. Obviously typeid(class) and typeid(struct) doesn't work. But...

1. Is this going to support typeid(C) where C is a class or struct type?
2. The message isn't clear enough that it supports class instance and struct instance typeid
3. I'm not sure why we need to restrict this to only classes and structs. Is there a technical reason why typeid(int) or typeid(i) (where i is an int) doesn't work?

Ideally `typeid(anything).name` works for everything the same as it does at runtime. Then code doesn't have to be written specifically for ctfe mode.

A further question I just thought of out of ignorance. Is the result of this going to be identical to typeid(x).name during runtime? If not, then it needs to be.

File: test/compilable/typeid_name.d (hunk 5 above)
Comment (Geod24) [reply test/compilable/typeid_name.d]: If the message is removed, then the test can be changed.

File: src/dmd/dinterpret.d (hunk 4 above)
Comment (stefan-koch-sociomantic) [reply src/dmd/dinterpret.d]: the reason typeid(basic_type).name is not supported is that it would take more code to do it.

File: src/dmd/dinterpret.d (hunk 1 above)
Comment (RazvanN7) [reply src/dmd/dinterpret.d]: The review is based on the provided code, not on promises of future commits. When `tie` will be used more than once only then using a variable for it is going to be justified, but for now -1 line of code is the preferred way.

File: src/dmd/dinterpret.d (hunk 2 above)
Comment (RazvanN7) [reply src/dmd/dinterpret.d]: The suggested change does not affect the nesting level, it simply limits the scope of `ident` to the block where it is used, getting rid of a line of code. This is the standard policy in such situations throughout the dmd codebase.

File: src/dmd/dinterpret.d (hunk 3 above)
Comment (RazvanN7) [reply src/dmd/dinterpret.d]: Simply grouping the instructions in a comment + a paragraph (preceded and succeed by a blank line) is more then enough from a readability standpoint and it uses less lines of code.

File: src/dmd/dinterpret.d (hunk 2 above)
Comment (UplinkCoder) [reply src/dmd/dinterpret.d]: ah yes can do.

File: src/dmd/dinterpret.d (hunk 4 above)
Comment (RazvanN7) [reply src/dmd/dinterpret.d]: Having an error message that is informative to the user is crucial to the acceptance of the PR.

File: test/compilable/typeid_name.d (hunk 5 above)
Comment (RazvanN7) [reply test/compilable/typeid_name.d]: We need coverage of all the lines of code introduced.

File: src/dmd/dinterpret.d (hunk 1 above)
Comment (UplinkCoder) [reply src/dmd/dinterpret.d]: There is one more reason to do it that way.
 `if (ex.op == TOK.typeid_)`
`cast(TypeIdExp)` should be close together since the former makes the later a valid operation.
These are extern(C++) classes which means that dynamic casts are not available.

Comment (RazvanN7) [reply src/dmd/dinterpret.d]: This is a valid argument only if `ex` is modified to point to a different type of class, but that is not happening here and it seems unlikely to happen because then you would rewrite the AST.

I am sure there a lot of situations where it makes sense to the the cast outside of the if branch, but the current situation is not one of them. For now, apply the required changes and when you will make the next PR extending the current code you can add the necessary modifications.

Comment (MoonlightSentinel) [reply src/dmd/dinterpret.d]: BTW you could also use `isTypeidExp` instead of explicitly checking `op`

Comment (UplinkCoder) [reply src/dmd/dinterpret.d]: which calls through the vtbl. why would you do that when you already had the op field?

Comment (Geod24) [reply src/dmd/dinterpret.d]: > which calls through the vtbl.

[Not since 2018](https://github.com/dlang/dmd/commit/edc7af85a29f67a6b5863e8338716beac03a9154)

Comment (UplinkCoder) [reply src/dmd/dinterpret.d]: ah yeah ... I don't usually develop on this version of dmd :)

**Review by RazvanN7 — APPROVED** [review]

**Comment by UplinkCoder** [comment]
why isn't it merged yet?

File: src/dmd/dinterpret.d [hunk 6]
```diff
@@ -6079,22 +6079,41 @@ public:
...
+
+        StructLiteralExp se;
+        int i;
+
+        if (ex.op != TOK.structLiteral && ex.op != TOK.classReference && ex.op != TOK.typeid_)
         {
+        LnotImplemented:
             e.error("`%s.%s` is not yet implemented at compile time", e.e1.toChars(), e.var.toChars());
             result = CTFEExp.cantexp;
             return;
```
Comment (thewilsonator) [inline src/dmd/dinterpret.d]: Please make this a nested function so that no gotos are needed.

File: src/dmd/dinterpret.d [hunk 7]
```diff
@@ -6079,22 +6079,41 @@ public:
...
+                    auto sym = t.toDsymbol(null);
+                    if (auto ident = (sym ? sym.ident : null))
+                    {
+                        result = new StringExp(e.loc, ident.toString());
+                        result.expressionSemantic(null);
+                        return ;
+                    }
+                }
+            }
+            goto LnotImplemented;
```
Comment (thewilsonator) [inline src/dmd/dinterpret.d]: and call it here

**Comment by Geod24** [comment]
> why isn't it merged yet?

As far as I can see, some of the feedback hasn't been addressed (e.g. error message).
Also the `goto` thing that Nicholas just mentioned.

**Comment by UplinkCoder** [comment]
Now the error message does not need a separate test since it's already
tested.
